File: lib/java/openlinktoken/src/main/java/org/openlinktoken/tokens/tokenizer/SHAKE256TokenDigest.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.tokens.tokenizer;

import org.bouncycastle.crypto.digests.SHAKEDigest;

/**
 * Calculates fixed-width SHAKE256 token digests.
 */
public final class SHAKE256TokenDigest implements TokenDigest {
    private static final int OUTPUT_LENGTH_BYTES = 32;

    /**
     * Calculates the first 32 bytes of a SHAKE256 digest.
     *
     * @param value the token-signature bytes
     * @return a 32-byte SHAKE256 digest
     */
    @Override
    public byte[] digest(byte[] value) {
        SHAKEDigest digest = new SHAKEDigest(256);
        digest.update(value, 0, value.length);
        byte[] output = new byte[OUTPUT_LENGTH_BYTES];
        digest.doFinal(output, 0, output.length);
        return output;
    }
}

File: lib/python/openlinktoken/src/main/openlinktoken/tokens/tokenizer/shake256_token_digest.py
# SPDX-License-Identifier: MIT
"""SHAKE256 token digest implementation."""

import hashlib

from openlinktoken.tokens.tokenizer.token_digest import TokenDigest


class Shake256TokenDigest(TokenDigest):
    """Calculate a fixed-width SHAKE256 token digest."""

    OUTPUT_LENGTH = 32

    def digest(self, value: bytes) -> bytes:
        """Return 32 bytes from the SHAKE256 extendable-output function."""
        return hashlib.shake_256(value).digest(self.OUTPUT_LENGTH)
